mountpoint-s3-client/src/s3_crt_client/get_object.rs

```diff
@@ -109,16 +134,32 @@ pub struct S3GetObjectRequest {
     finish_receiver: UnboundedReceiver<Result<GetBodyPart, Error>>,
     finished: bool,
     enable_backpressure: bool,
+    object_metadata_cb: Lazy<Result<ObjectMetadata, Canceled>, Receiver<ObjectMetadata>>,
     /// Next offset of the data to be polled from [poll_next]
     next_offset: u64,
     /// Upper bound of the current read window. When backpressure is enabled, [S3GetObjectRequest]
     /// can return data up to this offset *exclusively*.
     read_window_end_offset: u64,
 }
 
+#[async_trait]
 impl GetObjectRequest for S3GetObjectRequest {
     type ClientError = S3RequestError;
 
+    async fn get_object_metadata(self: Pin<&Self>) -> Result<ObjectMetadata, Self::ClientError> {
+        let empty_read_window = self.read_window_end_offset <= self.next_offset;
+        let request_required = self.object_metadata_cb.try_get().is_none();
+        if self.enable_backpressure && empty_read_window && request_required {
```

Comment:
I propose to limit this to the `initial_read_window == 0` case, so the check will be as follows:
```rust
let request_required = self.object_metadata_cb.try_get().is_none();
if request_required && self.enable_backpressure && self.initial_read_window == 0 {
    return Err(S3RequestError::EmptyReadWindow);
}
```

(we'll need to store `initial_read_window ` in the request [on creation](https://github.com/awslabs/mountpoint-s3/blob/56dd238c43f8219ede7f449f92bd2632798a87aa/mountpoint-s3-client/src/s3_crt_client/get_object.rs#L71) of it)